src/mixins/labels.js

```diff
@@ -0,0 +1,115 @@
+import * as THREE from "three";
+
+import { ATOM_GROUP_NAME } from "../enums";
+
+/*
+ * Mixin containing the logic for dealing with atom labes.
+ * Dynamically draws labels over atoms.
+ */
+export const LabelsMixin = (superclass) =>
+    class extends superclass {
+        // eslint-disable-next-line class-methods-use-this
+        createTextTexture(text) {
+            const {
+                fontFace,
+                fontSize,
+                fontWeight,
+                fillColor,
+                strokeColor,
+                strokeLineWidth,
+                textAlign,
+                textBaseline,
+            } = this.settings.labels;
+            const canvas = document.createElement("canvas");
+            const context = canvas.getContext("2d");
+            context.font = `${fontWeight} ${fontSize}px ${fontFace}`;
+            const metrics = context.measureText(text);
+            const textWidth = metrics.width;
+            const basicTextSize = textWidth > fontSize ? textWidth : fontSize;
+            const textSizePowOf2 = 2 ** Math.floor(Math.log2(basicTextSize));
+            const scaledFontSize = (fontSize * textSizePowOf2) / basicTextSize;
+            canvas.width = textSizePowOf2;
+            canvas.height = textSizePowOf2;
+            context.font = `${fontWeight} ${scaledFontSize}px ${fontFace}`;
+            context.textAlign = textAlign;
+            context.textBaseline = textBaseline;
+            context.fillStyle = fillColor;
+            context.strokeStyle = strokeColor;
+            context.lineWidth = strokeLineWidth;
+            context.fillText(text, context.canvas.width / 2, (context.canvas.height / 2) * 1.15);
+            context.strokeText(text, context.canvas.width / 2, (context.canvas.height / 2) * 1.15);
+            const texture = new THREE.Texture(canvas);
+            texture.needsUpdate = true;
+            return texture;
+        }
+
+        createSprite(text, name) {
+            const texture = this.createTextTexture(text);
+            const spriteMaterial = new THREE.SpriteMaterial({ map: texture });
+            const sprite = new THREE.Sprite(spriteMaterial);
+            sprite.name = name;
+            sprite.scale.set(0.25, 0.25, 0.25);
+            sprite.visible = this.areLabelsShown;
+            return sprite;
+        }
+
+        adjustLabelsToCameraPosition() {
```

Comment:
It looks like this method gets called every time the `render` function is called. I think an independent loop over atoms could be expensive for a scene with many atoms, especially if it were being rendered as a "trajectory" or time series of atomic positions. We likely wouldn't need labels rendered if "watching a movie" anyway, so maybe we can just add a check `if (!this.areLabelsShown) { return }`

Reply:
Thanks! Agreed.